[CALCITE-5918] Add MAP function (enabled in Spark library)

chucheng92 opened the pull request:
https://issues.apache.org/jira/browse/CALCITE-5918

the spark map function has 2 different points with calcite map constructor:
1. use map(k, v, ...) rather than map[k, v, ...]
2. allows empty map such as map(); however calcite map constructor not support it.

macroguo-ghy commented on core/src/main/java/org/apache/calcite/sql/fun/SqlLibraryOperators.java:
```diff
@@ -1077,6 +1079,38 @@ private static RelDataType arrayReturnType(SqlOperatorBinding opBinding) {
+      List<RelDataType> argTypes) {
+    // special case, allows empty map
+    if (argTypes.size() == 0) {
+      return Pair.of(typeFactory.createUnknownType(), typeFactory.createUnknownType());
```
There is a subtle difference between `unknown` and `any`, can you please explain why `unknown` is used?

chucheng92 replied:
because Spark use UNKNOWN type when it's empty, it's similar with current spark ARRAY handling. pls see: https://github.com/apache/calcite/pull/3141/files

chucheng92 commented on site/_docs/reference.md:
```diff
@@ -2769,6 +2769,7 @@ BigQuery's type system uses confusingly different names for types and functions:
 | b | TO_HEX(binary)                                 | Converts *binary* into a hexadecimal varchar
 | b | FROM_HEX(varchar)                              | Converts a hexadecimal-encoded *varchar* into bytes
 | b o | LTRIM(string)                                | Returns *string* with all blanks removed from the start
+| s | MAP(key, value [, key, value]*)                | Returns a map with the given key/value pairs
```
yes, good point. I will fix it later.

chucheng92 replied:
Fixed.

tanclary replied:
Also when referencing the function arguments please surround them with * to remain consistent with other functions

chucheng92 replied:
yes, fixed.

chucheng92 commented on core/src/main/codegen/templates/Parser.jj:
```diff
@@ -4886,14 +4886,21 @@ SqlNode MapConstructor() :
+                return SqlStdOperatorTable.MAP_QUERY.createCall(s.end(this), args.get(0));
+            } else {
+                // MAP function e.g. "MAP(1, 2)" equivalent to standard "MAP[1, 2]"
+                return SqlLibraryOperators.MAP.createCall(s.end(this), args.getList());
```
I hope to retain the original parser logic here, because it will be detected later on OperandTypeChecker. And here is similar with Array handling. pls see: https://github.com/apache/calcite/pull/3141/files

chucheng92 commented on testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java:
```diff
@@ -6604,11 +6605,44 @@ private static void checkIf(SqlOperatorFixture f) {
     f.checkFails("^map_concat(map[1, null], array[1])^",
         "Parameters must be of the same type", false);
+
+    // 2. check with map function, map(k, v ...)
```
yes, good point. I will fix it later.

chucheng92 replied:
updated, but currently calcite not support 'cast(map() as map<string, int>)', so I have added the case in IF block until we support it.

chucheng92 commented:
@macroguo-ghy hi, thanks for reviewing. 
Because we have implemented Spark's ARRAY function, their processing logic on parser and empty array/map is similar, if compare with 2 PRs to review may be more clearer.

https://github.com/apache/calcite/pull/3141/files

tanclary commented on core/src/main/java/org/apache/calcite/sql/fun/SqlLibraryOperators.java:
```diff
@@ -1082,6 +1084,38 @@ private static RelDataType arrayReturnType(SqlOperatorBinding opBinding) {
+      RelDataTypeFactory typeFactory,
+      List<RelDataType> argTypes) {
+    // special case, allows empty map
+    if (argTypes.size() == 0) {
```
should we use `isEmpty()` instead?

chucheng92 replied:
yes, fixed

tanclary commented on core/src/main/java/org/apache/calcite/sql/type/OperandTypes.java:
```diff
@@ -568,6 +569,9 @@ public static SqlOperandTypeChecker variadic(
   public static final SqlSingleOperandTypeChecker MAP_FROM_ENTRIES =
       new MapFromEntriesOperandTypeChecker();
 
+  public static final SqlSingleOperandTypeChecker MAP_FUNCTION =
```
same naming comment as above

chucheng92 replied:
fixed.

tanclary commented on core/src/main/java/org/apache/calcite/sql/type/OperandTypes.java:
```diff
@@ -1221,6 +1225,53 @@ private static class MapFromEntriesOperandTypeChecker
+      extends SameOperandTypeChecker {
+
+    MapFunctionOperandTypeChecker() {
+      super(-1);
```
where does the -1 come from?

tanclary commented on core/src/main/java/org/apache/calcite/sql/type/OperandTypes.java:
```diff
@@ -1221,6 +1225,53 @@ private static class MapFromEntriesOperandTypeChecker
+      final List<RelDataType> argTypes =
+          SqlTypeUtil.deriveType(callBinding, callBinding.operands());
+      // allows empty map
+      if (argTypes.size() == 0) {
```
`isEmpty()`

tanclary commented on core/src/main/java/org/apache/calcite/sql/type/OperandTypes.java:
```diff
@@ -1221,6 +1225,53 @@ private static class MapFromEntriesOperandTypeChecker
+        return true;
+      }
+      // the size of map arg types must be even.
+      if (argTypes.size() % 2 > 0) {
```
is the condition more clear if we say `argTypes.size() % 2 != 0`

chucheng92 replied:
fixed.

tanclary commented on core/src/main/java/org/apache/calcite/sql/type/OperandTypes.java:
```diff
@@ -1221,6 +1225,53 @@ private static class MapFromEntriesOperandTypeChecker
+        RelDataTypeFactory typeFactory,
+        List<RelDataType> argTypes) {
+      return Pair.of(
+          typeFactory.leastRestrictive(Util.quotientList(argTypes, 2, 0)),
```
Where are these constants coming from and what do they signify?

chucheng92 replied:
Util.quotientList(argTypes, 2, 0):
This extracts all elements at even indices from argTypes.
It represents the types of keys in the map as they are placed at even positions
e.g. 0, 2, 4, etc. details please see Util.quotientList.

I've added some comments here for better readability.

JiajunBernoulli replied:
Can we add UT for `typeFactory.leastRestrictive`?

How about?
`MAP('k1',  1, 'k2', 2.0, 'k3', '3')`

chucheng92 replied:
> Can we add UT for `typeFactory.leastRestrictive`?
> 
> How about? `MAP('k1', 1, 'k2', 2.0, 'k3', '3')`

yes, added it. thanks for reviewing.

tanclary commented on site/_docs/reference.md:
```diff
@@ -2769,6 +2769,8 @@ BigQuery's type system uses confusingly different names for types and functions:
 | b | TO_HEX(binary)                                 | Converts *binary* into a hexadecimal varchar
 | b | FROM_HEX(varchar)                              | Converts a hexadecimal-encoded *varchar* into bytes
 | b o | LTRIM(string)                                | Returns *string* with all blanks removed from the start
+| s | MAP()                                          | Returns a empty map
```
nit: an

tanclary commented on testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java:
```diff
@@ -6703,11 +6704,53 @@ private static void checkIf(SqlOperatorFixture f) {
+
+    // after calcite supports cast(null as map<string, int>), cast(map() as map<string, int>)
+    // it should add these tests.
+    if (TODO) {
```
instead of `if (TODO)` would it be better to add an enum to the `Bug` class? Here is an example of what I mean: https://github.com/apache/calcite/blob/main/testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java#L624

chucheng92 replied:
I get your point. but the form such as `cast(map() as map<string, int>)` may not necessarily be a bug, and there is currently no JIRA to track it, calcite is not ready to support it yet, and may not support it, so it is kept as a TODO for the time being.

chucheng92 commented on core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java:
```diff
@@ -5251,6 +5251,17 @@ public static Map mapFromArrays(List keysArray, List valuesArray) {
+  public static Map mapFunction(Object... args) {
+    final Map map = new LinkedHashMap<>();
+    for (int i = 0; i < args.length; i++) {
+      Object key = args[i++];
```
yes, updated.

chucheng92 commented on core/src/main/java/org/apache/calcite/adapter/enumerable/RexImpTable.java:
```diff
@@ -873,6 +874,7 @@ Builder populate2() {
       map.put(MAP_VALUE_CONSTRUCTOR, value);
       map.put(ARRAY_VALUE_CONSTRUCTOR, value);
       defineMethod(ARRAY, BuiltInMethod.ARRAYS_AS_LIST.method, NullPolicy.NONE);
+      defineMethod(MAP, BuiltInMethod.MAP_FUNCTION.method, NullPolicy.NONE);
```
In fact, there is a MAP object in BuiltInMethod, occupy this name. but yes, if we rename that, we can use MAP. I have updated it. WDYT?

chucheng92 commented on core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java:
```diff
@@ -5251,6 +5251,17 @@ public static Map mapFromArrays(List keysArray, List valuesArray) {
   }
 
+  /** Support the MAP function. */
+  public static Map mapFunction(Object... args) {
```
fixed.

chucheng92 commented:
hi, @tanclary I have addressed all your comments except 
`instead of if (TODO) would it be better to add an enum to the Bug class?` this one.

I have rebased the main accidentally, I hope it won’t cause any trouble to your review. 

If you have time, pls help me to review it ?

chucheng92 commented:
I have rebased to the main to use [CALCITE-5570](https://issues.apache.org/jira/browse/CALCITE-5570), it supports the form like 'cast(null as map<int, int>)'

chucheng92 commented:
hi, @tanclary Sorry to pin you, I have solved the comments. could you help to re-check it?

chucheng92 commented on core/src/main/java/org/apache/calcite/prepare/LixToRelTranslator.java:
```diff
@@ -97,7 +97,7 @@ public <T> RelNode translate(Queryable<T> queryable) {
     if (expression instanceof MethodCallExpression) {
       final MethodCallExpression call = (MethodCallExpression) expression;
-      BuiltInMethod method = BuiltInMethod.MAP.get(call.method);
+      BuiltInMethod method = BuiltInMethod.FUNCTIONS_MAPS.get(call.method);
```
because this is a static field in BuiltInMethod to store all the functions. it use 'MAP' name before, so I can't continue to use this name as our spark map method name as you suggested. (I set MAP_FUNCTION previously).  

one disadvantage I can think of here is that it may cause compatibility issues? If there is an upper-layer application that reflects the use of this field. 

or should i reset spark MAP function to 'MAP_FUNCTION' name? what do you think?

tanclary reviewed (commented):
I think this looks good, you might find some of the sonarcloud reports helpful, it looks like they have some suggestions on refactoring the `for` loops and providing types for the generic `Map`.

chucheng92 commented:
@tanclary 

Thanks for the reminder. I checked sonar report and problems are roughly divided into three categories:
1. Assignment of index i in for-loop
2. The MAP of SqlFunctions does not specify a generic type
3. Reuse string constants

I have modified 1, but 2 and 3 actually not applicable for us. 

2: The methods of SqlFunctions are all used by reflection. It can be written as generics or not. However, the current collection types Map and List are expressed using raw, and they tend to be consistent with them.

3: is something like "CHAR(10) NOT NULL", there are multiple checkScalar, sonar wants us to define a variable and reuse it. In fact, it is not necessary. The current way of writing is more intuitive.

If you agree with these, I will squash the commits. WDYT?

tanclary reviewed (approved):
LGTM, think there might be some checkstyle issues but feel free to squash and ping me after!

chucheng92 commented:
thank you all for patient reviewing! all comments are resolved. ci passed.

@tanclary could you help to merge it? so we can add it in up-coming 1.36.

tanclary reviewed (approved):
LGTM, sorry for the delay! Congrats on becoming a committer!

chucheng92 commented:
> LGTM, sorry for the delay! Congrats on becoming a committer!

thanks tanner. you helped me to a lot. :)
